ocaml/xapi/message_forwarding.ml

```diff
@@ -1727,7 +1743,23 @@ module Forward = functor(Local: Custom_actions.CUSTOM_ACTIONS) -> struct
                assert_can_migrate ~__context ~vm ~dest ~live ~vdi_map ~vif_map ~vgpu_map ~options
              );
            forwarder ~local_fn ~__context ~vm
-             (fun session_id rpc -> Client.VM.migrate_send rpc session_id vm dest live vdi_map vif_map options vgpu_map)
+             (fun session_id rpc ->
+               let t = Client.Async.VM.migrate_send rpc session_id vm dest live vdi_map vif_map options vgpu_map in
+               D.info "VM.migrate_send waiting for migration to complete... on task %s" (Ref.string_of t);
+               wait_for_tasks ~__debug:true ~__context ~tasks:[t];
+               match rpc_of_task ~__context ~self:t with
+               | `Ok x -> begin
+                 try
+                  API.ref_VM_of_rpc x
+                 with e ->
+                   failwith @@ Printf.sprintf "impossible: VM ref was not placed in task's (ref = %s) result field. error: %s" (Ref.string_of t) (Printexc.to_string e)
+                 end
+               | `Cancelling | `Cancelled -> raise (Api_errors.Server_error (Api_errors.task_cancelled, [ Ref.string_of t]))
+               | `Server_error e -> raise e
+               | `Could_not_parse_xml_rpc e -> raise e
+               | `Unknown_error -> failwith "impossible: couldn't extract error info from task"
+               | `Pending ->  failwith "impossible: just waited on task"
+               | `Null -> failwith "impossible: vm ref wasn't written to task result")
```

Comment:
using `Client.Async.VM` instead of `Client.VM` introduces a lot of new error conditions (i hope most of these are actually impossible)

I'm not sure whether to introduce a new API error or whether there are suitable existing ones (there seem to be no generic `MIGRATION_FAIL` errors)

Reply:
Should these not be mapped to internal error if we expect them to never occur? You could use:

```
let fail msg  = Api_errors.(Server_error (internal_error, [msg])) 
```